subprojects/composite-builds/src/main/java/org/gradle/composite/internal/CompositeBuildClassPathInitializer.java

```diff
@@ -46,13 +60,29 @@ public void execute(Configuration classpath) {
             tasksToBuild.add(TaskIdentifier.of(targetBuild.getBuildIdentifier(), (TaskInternal) task));
         }
         if (!tasksToBuild.isEmpty()) {
-            buildTreeWorkGraphController.withNewWorkGraph(graph -> {
-                graph
-                    .scheduleWork(builder -> builder.scheduleTasks(tasksToBuild))
-                    .runWork()
-                    .rethrow();
-                return null;
-            });
+            buildTreeWorkGraphController.withNewWorkGraph(graph -> buildBuildLogic(graph, tasksToBuild));
+        }
+    }
+
+    @Nullable
+    @SuppressWarnings("try")
+    private Void buildBuildLogic(BuildTreeWorkGraph graph, List<TaskIdentifier.TaskBasedTaskIdentifier> tasksToBuild) {
+        BuildTreeWorkGraph.FinalizedGraph finalizedGraph = graph.scheduleWork(builder -> builder.scheduleTasks(tasksToBuild));
+        try (FileLock ignored = buildLogicBuildLockFor(currentBuild)) {
```

Comment:
💭 The lock here is only around building the tasks for the classpath. I suppose we'd need the also around the part where copy the resulting JARs in the Gradle caches? Though this lock should at least make sure that the build logic projects and their task state/build cache entries are correct.